### ISSUE TITLE
Support for setting balance

### ISSUE THREAD

**Issue by tephlon** [SoCo/SoCo#692]
I needed to be able to control the speaker balance, so I monkey patched support for it in my own script. I basically copied the volume getter/setter in core and change the volumes of the LF and RF channels. Works nicely on my paired Sonos Ones. Here is the gist of it, if anyone is interested.

```
def get_balance(self):
    """Get the speaker's balance.

    @return: A 2-tuple with integers between 0 and 100.
    """

    response_lf = self.renderingControl.GetVolume([
	('InstanceID', 0),
	('Channel', 'LF'),
    ])
    response_rf = self.renderingControl.GetVolume([
	('InstanceID', 0),
	('Channel', 'RF'),
    ])
    volume_lf = response_lf['CurrentVolume']
    volume_rf = response_rf['CurrentVolume']
    return int(volume_lf), int(volume_rf)

def set_balance(self, left, right):
    """Set the balance."""
    left = int(left)
    right = int(right)
    left = max(0, min(left, 100))  # Coerce in range
    right = max(0, min(right, 100))  # Coerce in range
    self.renderingControl.SetVolume([
	('InstanceID', 0),
	('Channel', 'LF'),
	('DesiredVolume', left)
    ])
    self.renderingControl.SetVolume([
	('InstanceID', 0),
	('Channel', 'RF'),
	('DesiredVolume', right)
    ])
```

Inspiration by https://github.com/jishi/node-sonos-http-api/issues/454 and https://github.com/chregu/fhem-sonos/blob/b6a97be8015b70644d71d74bf8224dcb20833a75/FHEM/00_SONOS.pm#L1207

**Comment by pwt** [comment]
Good work -- this is useful and personally I'd like to see it included in SoCo.

Are you going to put up a PR? If not, I'm happy to do so on your behalf. I've created a commit of the code above, with normalised docstrings at: https://github.com/pwt/SoCo/commit/ee671361768f12fe23124a524840c6ea8b1acaa4.

Feel free to grab this. It passes pep8, flake8, and the docstrings are consistent with the rest of SoCo.

**Comment by tephlon** [comment]
That's great! Please take over, I don't have the time :)



Den tors 28 nov. 2019 12:24pwt <notifications@github.com> skrev:

**Comment by pwt** [comment]
Ok, great! I'll look at adding some unit tests then pushing it as a PR. I'll credit you. (I've already functionally tested the code against my Sonos systems and it works perfectly.)

**Comment by tephlon** [comment]
Awesome! Thank you!




Den tors 28 nov. 2019 12:38pwt <notifications@github.com> skrev:

**Comment by pwt** [comment]
The PR (#693) is up. I changed the implementation from method calls to operations on a `balance` property, to make it more consistent with the rest of SoCo. I added unit tests and normalised the docstrings.

**Comment by tephlon** [comment]
Thanks a lot! :) Yeah, I just monkey patched the instance, and shortcut the whole omg-I-don't-know-if-the-property-and-setter-decorators-work-when-monkey-patching discovery process by using a plain old getter and setter :)

**Comment by KennethNielsen** [comment]
@tephlon thanks a lot for forwarding the information and thanks to @pwt for putting it into a PR. I will have a look at it as soon as I can.

**Comment by KennethNielsen** [comment]
Closed with #693

**Comment by tephlon** [comment]
Thanks a lot, @pwt and @KennethNielsen!
